### ISSUE TITLE
Don't require DLL to ship with app

### ISSUE THREAD

**Issue by nmoinvaz** [MicrosoftEdge/WebView2Feedback#110]
In issue #12, I mentioned:

> Shipping an additional DLL (WebView2Loader.dll) is problematic for us, as our [application] is only a single executable. In other platforms, Linux & Mac, and Windows with IE, it is not necessary to ship separate shared libraries, they are already installed on the system and we dynamically find the one to load and use.

> I think the best scenario would be to have the component installed on the system along-side Edge and possibly available as COM control through CoCreateInstance.

Due to some of the other problems presented in #12 that were solved, it has been closed. However this particular problem still remains and is similar to what was recently mentioned in #100.

> The current plan is for the end developer to ship that DLL with their app. We'll look into how we might avoid that. I'll open a task for that.

> For the BYO case Edge might not be installed so having a separate DLL is sort of lowest common denominator that works for all scenarios. But long term Edge should be installed and not requiring end developers ship a separate DLL would be much more convenient. I've got a task open to look into that.

It is possible to create a `IWebBrowser2` control without having to ship a separate DLL, we need some other similar method of instantiating a IWebView2 instance. 

If there is a way to access the same functions in the default Edge installation, perhaps via dynamically linking exports in `msedge.dll` or some other DLL that would even be an option.

[AB#25369096](https://microsoft.visualstudio.com/90b2a23c-cab8-4e7c-90e7-a977f32c1f5d/_workitems/edit/25369096)

**Comment by david-risney** [comment]
Yes, good point. Thanks for opening this issue.

**Comment by douglas-jordan** [comment]
Seems like they could change it to a link library in the SDK or at least make it an option.

**Comment by nmoinvaz** [comment]
Yeah a statically linked library would be good for native applications and a dynamically linked library would be good for .net applications.

**Comment by pabloko** [comment]
Faced this issue while creating a library plugin using WebView2 on a enviroment where i cannot put DLLs on executable's include path. I've worked around it by late-loading WebView2Loader.dll using LoadLibrary. Its some ugly hack but it will work, first, remove the reference to `WebView2Loader.dll.lib` in the `Microsoft.Web.WebView2.targets` file of the pkg, compiling now will throw a link error, then load the library:

```cpp
    FARPROC cwb2ewd;
...
//Call this early. entry point is perfect place
    char path[MAX_PATH] = { 0 };
    GetModuleFileName((HINSTANCE)&__ImageBase, path, _countof(path)); //Get this dll full path
    path[strlen(path) - 15] = '\0'; //My dll name is 15 chars long, so remove it
    sprintf(path, "%sWebView2Loader.dll", path); 
    HMODULE hMod = LoadLibrary(path); 
    if (hMod == NULL) {
      //exit here as cannot load the dll and will likely crash
    }
    cwb2ewd = GetProcAddress(hMod, "CreateWebView2EnvironmentWithDetails");
...
//Then call this method instead [CreateWebView2EnvironmentWithDetails] when creating your browser
HRESULT __stdcall __fake_CreateWebView2EnvironmentWithDetails(
  PCWSTR browserExecutableFolder,
  PCWSTR userDataFolder,
  PCWSTR additionalBrowserArguments,
  IWebView2CreateWebView2EnvironmentCompletedHandler* environment_created_handler) {
  return (*(HRESULT(__stdcall **)
    (PCWSTR, PCWSTR, PCWSTR, IWebView2CreateWebView2EnvironmentCompletedHandler*))&cwb2ewd)
    (browserExecutableFolder, userDataFolder, additionalBrowserArguments, environment_created_handler);
}
```
So, WebView2Loader.dll wont be present on the AIT and will let you spawn the process without it present.

**Comment by marcocantu** [comment]
I agree avoiding the DLL would be really much better for anyone deploying a native single EXE solution

**Comment by finalhit** [comment]
Any update on providing statically linked library?

Also wondering is WebView2Loader.dll present in system folder when Microsoft Edge Chromium is installed?

**Comment by marcocantu** [comment]
Agreed, it would be good to have an update. We are about to ship a product that integrates WebView2...

**Comment by david-risney** [comment]
We're still looking into a static lib and will let you know when we have more info.

There's no plan currently to have the loader dll in the system folder.

**Comment by nmoinvaz** [comment]
What would be the minimum OS requirements for the static library? We have an app that targets lower versions of Windows (XP, Vista, 7). If the static library uses functions only available in newer versions of Windows, then our app won't be able to run on older versions of Windows. I think it would be preferable to have a loader DLL in the system folder or somewhere else that we can dynamically load. If it exists, we can load it, if not we can load old-school _IWebBrowser_ engine.

**Comment by Ciantic** [comment]
> There's no plan currently to have the loader dll in the system folder.

Have you thought about having it in different folder, e.g. `%appdata%\..\Local\Microsoft\Edge` ?

**Comment by Urmeli0815** [comment]
We would also be interested in a library that can be statically linked. It reduces the effort as we don't need to distribute the loader DLL. Besides that having it as a DLL increases the attack surface, so additional steps are necessary to verify that it wasn't tampered with (certificate check etc).

**Comment by champnic** [comment]
Quick update: Most of the work for a library that can be statically linked is complete. It will likely be available in the next release, and we are working on documentation now. I'll update this thread again when it gets released. Thanks!

**Comment by nmoinvaz** [comment]
What are the minimum OS requirements for statically linked library?

**Comment by hassandraga** [comment]
Interesting, any new update?

**Comment by nmoinvaz** [comment]
It looks like the latest Nuget package includes the static libraries. Not sure about the minimum requirements.

**Comment by hassandraga** [comment]
Awesome news!, I'm using Microsoft.Web.WebView2 [_0.9.628-prerelease_], however I didn't found _Referenced Packages_ in _Project->Properties->Configuration Properties_.

How to make our project use a static linking for WebView2 please?

**Comment by hassandraga** [comment]
Changing Microsoft.Web.WebView2.targets to Static will cause a undefined reference., also I found WebView2LoaderStatic.lib only for ARM x64 Native.

**Comment by LiangTheDev** [comment]
Add the following in the VS project file should link to static lib instead of dll:

<PropertyGroup> <WebView2LoaderPreference>Alternate</WebView2LoaderPreference> </PropertyGroup>

WebView2LoaderStatic.lib should be there for all architectures (x86, x64 and arm64).

**Comment by douglas-jordan** [comment]
> Add the following in the VS project file should link to static lib instead of dll:
> 
> `<PropertyGroup> <WebView2LoaderPreference>Alternate</WebView2LoaderPreference> </PropertyGroup>`
> 
> WebView2LoaderStatic.lib should be there for all architectures (x86, x64 and arm64).

I made the change and the executable still has a webview2loader.dll reference in the dependencies.
Never made changes like this to the project file, so I am in the weeds a bit.  A complete description of the steps needed would be helpful.

**Comment by LiangTheDev** [comment]
I was actually might be wrong, please switch back to Static. What undefined references are you seeing?

**Comment by dianaqu** [comment]
Hi @douglas-jordan, @hassandraga,

Please make sure to include the following in your project files below the lines where you include our NuGet package targets files.
`<PropertyGroup> <WebView2LoaderPreference>Static</WebView2LoaderPreference> </PropertyGroup>`

So for example:
![image](https://user-images.githubusercontent.com/25068886/93150386-d1ed8200-f6ad-11ea-98af-f4b92114d195.png)

The WebView2LoaderPreference property in your project need to override the property in our targets file+

**Comment by douglas-jordan** [comment]
I set to Static and now I get unresolved externals, suspect some kind of unicode setting issue.

`
1>WebView2LoaderStatic.lib(loader_impl.obj) : error LNK2001: unresolved external symbol GetFileVersionInfoSizeW
1>WebView2LoaderStatic.lib(loader_impl.obj) : error LNK2001: unresolved external symbol GetFileVersionInfoW
1>WebView2LoaderStatic.lib(loader_impl.obj) : error LNK2001: unresolved external symbol VerQueryValueW`

**Comment by douglas-jordan** [comment]
I got it working, needed to just add version.lib as an additional dependency (with the propertygroup static change from @dianaqu ).  The webview2loader.dll is no longer listed in my exe as a dependency and confirmed working on Windows 10 with 18 separate pages.
I am using the v0.9.628-prerelease, Canary build 87.0.640.0 and VS2019.

**Comment by nmoinvaz** [comment]
This is what I am using for cmake. If there is a better way please let me know:

```
set(EDGE_RUNTIME_VERSION "2.0.9.622.11")
...
if(CMAKE_VERSION VERSION_LESS "3.15")
    message(FATAL "CMake 3.15 required for nuget package support")
endif()

set_target_properties(${target} PROPERTIES 
    VS_PACKAGE_REFERENCES "Microsoft.Web.WebView2_${EDGE_RUNTIME_VERSION}")

set(EDGE_PACKAGE_DIR ${CMAKE_CURRENT_SOURCE_DIR}/packages/Microsoft.Web.WebView${EDGE_RUNTIME_VERSION})
set(EDGE_INCLUDE_DIR ${EDGE_PACKAGE_DIR}/build/native/include)
set(EDGE_PLATFORM "x86")
if(CMAKE_GENERATOR_PLATFORM STREQUAL "x64")
    set(EDGE_PLATFORM "x64")
endif()

set(EDGE_PLATFORM_DIR ${EDGE_PACKAGE_DIR}/build/native/${EDGE_PLATFORM})

target_compile_definitions(${PROJECT_NAME} PRIVATE HAVE_EDGE)
target_include_directories(${PROJECT_NAME} PRIVATE ${EDGE_INCLUDE_DIR})
target_link_libraries(${PROJECT_NAME} ${EDGE_PLATFORM_DIR}/WebView2LoaderStatic.lib)
```

**Comment by champnic** [comment]
This should be fixed in [WebView2 SDK 1.0.674-prerelease](https://www.nuget.org/packages/Microsoft.Web.WebView2/1.0.674-prerelease). We've also added the necessary version.lib reference by default. If this is still an issue please let us know. Thanks!

**Comment by DineshSolanki** [comment]
For anyone still wondering why the `<PropertyGroup> <WebView2LoaderPreference>Static</WebView2LoaderPreference> </PropertyGroup>` is not working, 
you need to add the following to your publish profile as well - 

```csharp
<IncludeNativeLibrariesForSelfExtract>true</IncludeNativeLibrariesForSelfExtract>
```
